Update socketcan_ctypes.py

mdjarrel opened the pull request:
Added filter support based on socketcan_native implementation and comments in issues #25.

hardbyte commented on can/interfaces/socketcan/socketcan_ctypes.py:
```diff
@@ -50,13 +51,56 @@ def __init__(self,
+                filter_data.append(can_filter['can_mask'])
+            res = libc.setsockopt(self.socket, SOL_CAN_RAW,
+                CAN_RAW_FILTER, struct.pack(can_filter_fmt, *filter_data),
+                4*2*len(kwargs['can_filters'])
```
`kwargs` isn't defined in this method, should just be `can_filters`

hardbyte replied:
Replace the magic `4*2` numbers with a variable

hardbyte commented on can/interfaces/socketcan/socketcan_ctypes.py:
```diff
@@ -50,13 +51,56 @@ def __init__(self,
+        if 'can_filters' in kwargs and len(kwargs['can_filters']) > 0:
+            log.debug("Creating a filtered can bus")
+            can_filter_fmt = "={}I".format(2 * len(kwargs['can_filters']))
+            filter_data = []
```
We can just call the `set_filters` method here and remove the duplicated code.

mdjarrel replied:
Agreed

hardbyte commented on can/interfaces/socketcan/socketcan_ctypes.py:
```diff
@@ -50,13 +51,56 @@ def __init__(self,
+        # Add any socket options such as can frame filters
+        if 'can_filters' in kwargs and len(kwargs['can_filters']) > 0:
+            log.debug("Creating a filtered can bus")
+            can_filter_fmt = "={}I".format(2 * len(kwargs['can_filters']))
```
I'm tempted to say this should be moved into `socketcan_constants.py` as it is shared with the native interface.

mdjarrel commented:
Thank you for the review @hardbyte . I have moved the code to convert the filter dictionary array to a sockopt arg array into socketcan_constants.py. I have also updated both the ctypes and native implementations to use the new function in socketcan_constants.py.
I also moved the code to take a None argument for can_filters and populate a wide open filter to the same function as the conversion code.
I am using the set_filters function now for setting filters in the constructor in socketcan_ctypes.py and fixed the reference to kwargs as well.
I also removed the magic 4*2 number and replaced it with a call to ctypes.sizeof(ctypes.c_uint32) and the length of the converted filter array instead of the filter dictionary array.
I changed my print call when setting the filters to use log.debug as well.
I quickly tested this with a vcan interface on Ubuntu in both python 2.7.12 and python 3.4.5 with ctypes in 2.7 and native in 3.4 and both seem to work.
This should be ready for another review.

hardbyte commented on can/interfaces/socketcan/socketcan_constants.py:
```diff
@@ -81,3 +81,18 @@
+  for can_filter in can_filters:
+    filter_data.append(can_filter['can_id'])
+    filter_data.append(can_filter['can_mask'])
+  return can_filter_fmt, filter_data
```
I think it would be better just return `struct.pack(can_filter_fmt, *filter_data)` - or am I missing something?

Slightly uncomfortable adding code to a constants file, perhaps add this in a new `socketcan_common.py` module.

hardbyte commented on can/interfaces/socketcan/socketcan_ctypes.py:
```diff
@@ -88,18 +77,13 @@ def set_filters(self, can_filters=None):
+                              len(filter_data)*ctypes.sizeof(ctypes.c_uint32)
+                             )
+        if res != 0:
+            log.debug('Setting filters failed: ' + str(res))
```
This should be a much higher priority log message (e.g. `log.error`).

I'm tempted to say we should also raise a `CanError` exception. Have you seen it fail, do you know when it might?

mdjarrel replied:
I have not seen it fail and I am not sure when it might.

mdjarrel commented:
I moved parseCanFilters out of socketcan_constants.py and into socketcan_common.py.
I am also now returning the filter struct from parseCanFilters.
Updated optlen param in socketcan_ctypes setsockopt to be len(filter_struct) .
I elevated the log.debug to log.error. I made a note to revisit the possibility of raising a CanError exception.

hardbyte commented:
Thanks for the contribution!
